r/R/marker_heatmap.R

```diff
@@ -0,0 +1,37 @@
+#' Generates a marker heatmap
+#'
+#' @param req
+#' @param data
+#'
+#' @return
+#' @export
+#'
+#' @examples
+runMarkerHeatmap <- function(req, data) {
+  nFeatures <- req$body$nGenes
+  data <- getClusters(req$body$type, req$body$config$resolution, data)
+
+  all_markers <- presto::wilcoxauc(data, assay = "data", seurat_assay = "RNA")
+  # Filtering out repeated genes to improve visualization, based on lowest p-value.
+  # We could also use fold change.
+  pvalueAggregate <- aggregate(all_markers$padj, by = list(all_markers$feature), FUN = min)
+  minp <- pvalueAggregate$x[match(all_markers$feature, pvalueAggregate$Group.1)]
+  all_markers$minp <- minp
+  all_markers <- subset(all_markers, all_markers$padj <= all_markers$minp)
+  all_markers <- all_markers[!duplicated(all_markers$feature), ]
+
+  nFeatures <- as.integer(30 / (as.integer(max(all_markers$group)) + 1))
+  all_markers <- all_markers %>%
+    group_by(group) %>%
+    arrange(padj) %>%
+    arrange(group) %>%
+    dplyr::filter(row_number() %in% c(1:nFeatures))
```

Comment:
```suggestion
  all_markers <- all_markers %>%
   group_by(group) %>%
    arrange(padj) %>%
    slice_head(n=5)
```
 I regularly see datasets with ~25ish clusters so just showing e.g. 1 gene per cluster would happen